Parsing for at-rule @font-feature-values

mdubet opened the pull request:
#### 4303ed343b77ef5e1fa84de8de8dbd0000cd79f1
<pre>
Parsing for at-rule @font-feature-values
<a href="https://bugs.webkit.org/show_bug.cgi?id=246120">https://bugs.webkit.org/show_bug.cgi?id=246120</a>
rdar://100830599

Reviewed by Myles C. Maxfield and Patrick Angle.

This implementation defines the @font-feature-values at-rule
and the six children font variation at-rules (@styleset, @stylistic, @character-variant,
@swash, @ornaments, @annotation), but those children at-rules are not exposed in CSSOM.

<a href="https://www.w3.org/TR/css-fonts-4/#font-feature-values-syntax">https://www.w3.org/TR/css-fonts-4/#font-feature-values-syntax</a>

* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/test_font_feature_values_parsing-expected.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/JSCSSRuleCustom.cpp:
(WebCore::toJSNewlyCreated):
* Source/WebCore/css/CSSFontFeatureValuesRule.cpp: Added.
(WebCore::CSSFontFeatureValuesRule::CSSFontFeatureValuesRule):
(WebCore::CSSFontFeatureValuesRule::cssText const):
(WebCore::CSSFontFeatureValuesRule::reattach):
(WebCore::CSSFontFeatureValuesBlockRule::CSSFontFeatureValuesBlockRule):
(WebCore::CSSFontFeatureValuesBlockRule::cssText const):
(WebCore::CSSFontFeatureValuesBlockRule::reattach):
* Source/WebCore/css/CSSFontFeatureValuesRule.h: Added.
* Source/WebCore/css/CSSFontFeatureValuesRule.idl: Added.
* Source/WebCore/css/StyleRule.cpp:
(WebCore::StyleRuleBase::visitDerived):
(WebCore::StyleRuleBase::createCSSOMWrapper const):
(WebCore::StyleRuleFontFeatureValues::StyleRuleFontFeatureValues):
(WebCore::StyleRuleFontFeatureValuesBlock::StyleRuleFontFeatureValuesBlock):
(WebCore::StyleRuleFontFeatureValues::create):
* Source/WebCore/css/StyleRule.h:
(WebCore::StyleRuleBase::isFontFeatureValuesRule const):
(WebCore::StyleRuleBase::isFontFeatureValuesBlockRule const):
(isType):
* Source/WebCore/css/StyleRuleType.h:
* Source/WebCore/css/StyleSheetContents.cpp:
(WebCore::StyleSheetContents::traverseSubresources const):
* Source/WebCore/css/parser/CSSAtRuleID.cpp:
(WebCore::cssAtRuleID):
* Source/WebCore/css/parser/CSSAtRuleID.h:
* Source/WebCore/css/parser/CSSParserImpl.cpp:
(WebCore::computeNewAllowedRules):
(WebCore::CSSParserImpl::consumeRuleList):
(WebCore::CSSParserImpl::consumeAtRule):
(WebCore::CSSParserImpl::consumeQualifiedRule):
(WebCore::fontFeatureValuesTypeMappings):
(WebCore::CSSParserImpl::consumeFontFeatureValuesRuleBlock):
(WebCore::CSSParserImpl::consumeFontFeatureValuesRule):
* Source/WebCore/css/parser/CSSParserImpl.h:
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::consumeFamilyNameList):
* Source/WebCore/css/parser/CSSPropertyParserHelpers.h:
* Source/WebCore/inspector/InspectorStyleSheet.cpp:
(flattenSourceData):

Canonical link: <a href="https://commits.webkit.org/255677@main">https://commits.webkit.org/255677@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/35479ad557477ae371f5550faac822bc18d8ab85

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/93294 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/2491 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/23943 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/102982 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/163267 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/97297 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/2498 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/30805 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/85671 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/99091 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/98964 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/1742 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/79755 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/28709 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/83655 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/83422 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/71771 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/84608 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/37192 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/17290 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/79674 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/35013 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/18537 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/27535 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3935 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/38887 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/41068 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/82308 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/40816 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/37757 "Passed tests") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/18597 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

patrickangle commented on Source/WebCore/css/parser/CSSParserImpl.cpp:
```diff
@@ -592,6 +602,138 @@ RefPtr<StyleRuleFontFace> CSSParserImpl::consumeFontFaceRule(CSSParserTokenRange
+    }
+}
+
+RefPtr<StyleRuleFontFeatureValuesBlock> CSSParserImpl::consumeFontFeatureValuesRuleBlock(CSSAtRuleID id, CSSParserTokenRange prelude, CSSParserTokenRange range)
```
Both of these new rule parsing paths should call into m_observableWrapper->observer() so that Web Inspector's parsing observer doesn't end up with an incorrect count of rules. See `consumeFontPaletteValuesRule` for an example of what that looks like.

patrickangle replied:
I guess technically we don’t keep these rules around, but we should still make sure the observer gets a complete picture.

mdubet replied:
I've tried to add support for the observer, but I would like to add a test to be sure that the Web Inspector is working properly. Do we have an example of such a test ?

mdubet replied:
@patrickangle I've added the observer code, if you want to take a look at it.

litherum commented on Source/WebCore/css/CSSFontFeatureValuesRule.cpp:
```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright (C) 2021 Apple Inc. All rights reserved.
```
2022

litherum commented on Source/WebCore/css/CSSFontFeatureValuesRule.cpp:
```diff
@@ -0,0 +1,107 @@
+            first = false;
+        }
+    };
+    joinFontFamiliesWithSeparator(m_fontFeatureValuesRule->fontFamilies(), ", "_s);
```
What's the point of the lambda if you're just going to immediately call it (just once)?

mdubet replied:
I personally find it cleaner, but I don't have a strong opinion on this.

litherum commented on Source/WebCore/css/CSSFontFeatureValuesRule.cpp:
```diff
@@ -0,0 +1,107 @@
+
+String CSSFontFeatureValuesBlockRule::cssText() const
+{
+    // This rule is always contained inside a FontFeatureValuesRule,
+    // which is the only one we are expected to serialize to CSS.
+    // We should never serialize a Block by itself.
```
Why not move the relevant code into here, and have FontFeatureValuesRule call this function?

mdubet replied:
Because it's not exactly the same: in the FontFeatureValuesRule, the duplicates are removed for example and the order is predefined (cf commit message). In general, this patch doesn't try to remember the CSS author exact syntax for the FontFeatureValuesBlockRule, but only the "effective" intent.

litherum commented on Source/WebCore/css/CSSFontFeatureValuesRule.h:
```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright (C) 2021 Apple Inc. All rights reserved.
```
2022

litherum commented on Source/WebCore/css/CSSFontFeatureValuesRule.h:
```diff
@@ -0,0 +1,73 @@
+    {
+        return m_fontFeatureValuesRule->fontFamilies();
+    }
+private:
```
Empty line before private, please

litherum commented on Source/WebCore/css/CSSFontFeatureValuesRule.h:
```diff
@@ -0,0 +1,73 @@
+public:
+    static Ref<CSSFontFeatureValuesRule> create(StyleRuleFontFeatureValues& rule, CSSStyleSheet* sheet) { return adoptRef(*new CSSFontFeatureValuesRule(rule, sheet)); }
+    virtual ~CSSFontFeatureValuesRule() = default;
+    const Vector<AtomString>& fontFamilies() const
```
I'd put an empty line above this, for cleanliness

litherum commented on Source/WebCore/css/CSSFontFeatureValuesRule.h:
```diff
@@ -0,0 +1,73 @@
+public:
+    static Ref<CSSFontFeatureValuesBlockRule> create(StyleRuleFontFeatureValuesBlock& rule, CSSStyleSheet* sheet) { return adoptRef(*new CSSFontFeatureValuesBlockRule(rule, sheet)); }
+    virtual ~CSSFontFeatureValuesBlockRule() = default;
+private:
```
ditto

litherum commented on Source/WebCore/css/StyleRule.h:
```diff
@@ -179,6 +182,69 @@ class StyleRuleFontPaletteValues final : public StyleRuleBase {
+    Tags annotation;
+};
+
+enum class FontFeatureValuesType {
+    Styleset,
+    Stylistic,
+    CharacterVariant,
+    Swash,
+    Ornaments,
+    Annotation
+};
```
Sweet! So cool.

litherum commented on Source/WebCore/css/StyleRule.h:
```diff
@@ -200,7 +266,6 @@ class StyleRulePage final : public StyleRuleBase {
     Ref<StyleProperties> m_properties;
     CSSSelectorList m_selectorList;
 };
-
```
Why?

mdubet replied:
Fixed

litherum commented on Source/WebCore/css/StyleRuleType.h:
```diff
@@ -41,6 +41,9 @@ enum class StyleRuleType : uint8_t {
     Supports = 12,
+    FontFeatureValues = 14,
+    // Those at-rules (@swash, @annotation,...) don't have a specified number in the spec.
+    FontFeatureValuesBlock = 15,
```
Should we use a random number, then, to avoid future collisions?

mdubet replied:
Maybe, what would you suggest ?

nt1m replied:
We should not use 15, it conflicts with an obsolete rule type (the `@viewport` one): https://w3c.github.io/csswg-drafts/cssom-1/#the-cssrule-interface

Any number above is fine, though, given the `type` is deprecated (e.g. meaningless). So I'd suggest `20`. If that's not already the case, we should make sure to always return `0` if it's not one of the defined types in the CSSOM spec, to be consistent with the spec. (that can be done in

litherum commented on Source/WebCore/css/parser/CSSParserImpl.cpp:
```diff
@@ -373,15 +377,22 @@ RefPtr<StyleRuleBase> CSSParserImpl::consumeAtRule(CSSParserTokenRange& range, A
     if (allowedRules == NoRules)
         return nullptr;
 
-    ASSERT(allowedRules <= RegularRules);
```
Why remove this?

mdubet replied:
This _ASSERT_ is a bit of a code smell imho. The previous lines "bypass" the switch statement with some early return and then just checks that we effectively bypassed it.
I could do the same (early function call + return) and keep the assert, but I think it's cleaner (less _if_ soup) to just remove this and flow normally (aka dispatch in the _switch_).

litherum commented on Source/WebCore/css/parser/CSSParserImpl.cpp:
```diff
@@ -417,7 +428,6 @@ RefPtr<StyleRuleBase> CSSParserImpl::consumeQualifiedRule(CSSParserTokenRange& r
     if (allowedRules == KeyframeRules)
         return consumeKeyframeStyleRule(prelude, block);
 
-    ASSERT_NOT_REACHED();
```
Ditto.

mdubet replied:
Same explanation (cf early function call + return in the previous lines).

litherum commented on Source/WebCore/css/parser/CSSParserImpl.cpp:
```diff
@@ -592,6 +602,138 @@ RefPtr<StyleRuleFontFace> CSSParserImpl::consumeFontFaceRule(CSSParserTokenRange
+{
+    switch (id) {
+    // Styleset has no maximum in theory, but in practise we don't consider more than 99 values.
+    case CSSAtRuleStyleset: return { FontFeatureValuesType::Styleset, 99 }; 
```
Should there be a fixme with a link to a new bug to avoid having to use an arbitrary number here?

mdubet replied:
I don't think we should try to support it because it will not be used by the platform.

I've added a link to the MDN sentence : "Note that values higher than 99 are valid, but don't map to any OpenType values and are ignored." https://developer.mozilla.org/en-US/docs/Web/CSS/@font-feature-values#syntax

litherum commented on Source/WebCore/css/parser/CSSParserImpl.cpp:
```diff
@@ -592,6 +602,138 @@ RefPtr<StyleRuleFontFace> CSSParserImpl::consumeFontFaceRule(CSSParserTokenRange
+
+RefPtr<StyleRuleFontFeatureValuesBlock> CSSParserImpl::consumeFontFeatureValuesRuleBlock(CSSAtRuleID id, CSSParserTokenRange prelude, CSSParserTokenRange range)
+{
+    // @swash { foo: 3; }
```
?

mdubet replied:
Changed.

litherum commented on Source/WebCore/css/parser/CSSParserImpl.cpp:
```diff
@@ -592,6 +602,138 @@ RefPtr<StyleRuleFontFace> CSSParserImpl::consumeFontFaceRule(CSSParserTokenRange
+{
+    // @swash { foo: 3; }
+
+    // Prelude should be empty.
```
Does this comment answer "why" rather than "what"?

mdubet replied:
I think so actually 😅  For me, "What?" answer would be "we return early if the prelude is at end" (which indeed is a fairly useless comment).

Which comment would you suggest ?

litherum replied:
🆗

litherum commented on Source/WebCore/css/parser/CSSParserImpl.cpp:
```diff
@@ -592,6 +602,138 @@ RefPtr<StyleRuleFontFace> CSSParserImpl::consumeFontFaceRule(CSSParserTokenRange
+    if (!prelude.atEnd())
+        return { };
+    
+    // Block should be present.
```
Ditto.

litherum commented on Source/WebCore/css/parser/CSSParserImpl.cpp:
```diff
@@ -592,6 +602,138 @@ RefPtr<StyleRuleFontFace> CSSParserImpl::consumeFontFaceRule(CSSParserTokenRange
+}
+RefPtr<StyleRuleFontFeatureValues> CSSParserImpl::consumeFontFeatureValuesRule(CSSParserTokenRange prelude, CSSParserTokenRange block)
+{
+    // @font-feature-values <family-name># { <declaration-list> }
```
Ditto

litherum commented on Source/WebCore/css/parser/CSSParserImpl.cpp:
```diff
@@ -592,6 +602,138 @@ RefPtr<StyleRuleFontFace> CSSParserImpl::consumeFontFaceRule(CSSParserTokenRange
+    });
+    rules.shrinkToFit();
+    
+    // Convert block rules to value (remove duplicate...etc)
```
Ditto

litherum commented on Source/WebCore/css/parser/CSSParserImpl.cpp:
```diff
@@ -592,6 +602,138 @@ RefPtr<StyleRuleFontFace> CSSParserImpl::consumeFontFaceRule(CSSParserTokenRange
+        if (!block)
+            continue;
+        if (!block->isFontFeatureValuesBlockRule())
+            continue;
```
Should this be an ASSERT()? (if yes, the downcast will do the assert for you and you can just delete these 2 lines)

litherum commented on Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
```diff
@@ -4534,6 +4534,19 @@ AtomString consumeFamilyNameRaw(CSSParserTokenRange& range)
     return concatenateFamilyName(range);
 }
 
+Vector<AtomString> consumeFamilyNameList(CSSParserTokenRange& range)
```
This seems a bit of a surprising addition. Why aren't we sharing code with the existing infrastructure to parse font-family?

mdubet replied:
`font-family` parses both `<family-name>` and `<generic-name>`.

Here, we exclusively parse `<family-name>` : https://drafts.csswg.org/css-fonts-4/#family-name-value

mdubet replied:
I've added a comment on the prototype in the header file.

litherum commented on Source/WebCore/css/CSSFontFeatureValuesRule.idl:
```diff
@@ -0,0 +1,27 @@
+    Exposed=Window
+] interface CSSFontFeatureValuesRule : CSSRule {
+    readonly attribute sequence<CSSOMString> fontFamilies;
+};
```
Huh. That makes this easy, I guess.

mdubet replied:
Indeed this patch doesn't try to expose the children at-rules.
Gecko did the same https://searchfox.org/mozilla-central/source/dom/webidl/CSSFontFeatureValuesRule.webidl

I've added a comment in the IDL file.

litherum replied:
Oh, is this contrary to the spec?

Can we get the spec changed if both us and Gecko are going to do this?

mdubet replied:
Indeed it would be worth discussing/editing the spec

litherum commented on Source/WebCore/css/CSSFontFeatureValuesRule.idl:
```diff
@@ -0,0 +1,27 @@
+ * Boston, MA 02110-1301, USA.
+ */
+
+typedef USVString CSSOMString;
```
It seems a shame that this has to be here. It might be worth commenting in https://bugs.webkit.org/show_bug.cgi?id=232548 indicating that this is yet one more place where we have to pointlessly duplicate typedefs.

litherum commented on LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/test_font_feature_values_parsing-expected.txt:
```diff
@@ -3,47 +3,126 @@ PASS basic parse tests - @font-feature-values;
-FAIL basic parse tests - @font-feature-values bongo { @styleset { abc: 1 2 3a } } assert_true: unexpected syntax error expected true got false
-FAIL basic parse tests - @font-feature-values bongo { @styleset { abc: 1 2 3, def: 1; } } assert_true: unexpected syntax error expected true got false
-FAIL basic parse tests - @font-feature-values bongo { @blah @styleset { abc: 1 2 3; } } assert_true: unexpected syntax error expected true got false
+PASS basic parse tests - @font-feature-values bongo {  }
```
Are there any tests that make sure removing the rule from the stylesheet (using the CSSOM) causes the right thing to happen? We should make sure such a test exists.

litherum commented on LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/test_font_feature_values_parsing-expected.txt:
```diff
@@ -3,47 +3,126 @@ PASS basic parse tests - @font-feature-values;
- } } assert_true: unexpected syntax error expected true got false
-FAIL basic parse tests - @font-feature-values bongo {
+ } }
+FAIL serialization check - @font-feature-values bongo {
```
FAIL sounds bad

mdubet replied:
We don't keep the order of the descriptors in this patch, which is the reason of the failure.
Maybe we should ? Or maybe we should modify the test ?

litherum replied:
Let's do either of those 2 things.

litherum commented on LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/test_font_feature_values_parsing-expected.txt:
```diff
@@ -3,47 +3,126 @@ PASS basic parse tests - @font-feature-values;
+ 5 6
+ 7;
+ } }
+FAIL serialization check - @font-feature-values bongo {
```
Ditto

mdubet replied:
Ditto.

nt1m commented on Source/WebCore/css/CSSFontFeatureValuesRule.idl:
```diff
@@ -0,0 +1,30 @@
+[
+    Exposed=Window
+] interface CSSFontFeatureValuesRule : CSSRule {
+    readonly attribute sequence<CSSOMString> fontFamilies;
```
Do we differ from Gecko and the spec intentionally here?

```
attribute CSSOMString fontFamily;
```

is what I see here. I assume it is an implementation detail?

mdubet replied:
https://www.w3.org/TR/css-fonts-4/#cssfontfeaturevaluesrule

I've changed it to be closer to the spec (even if I think it should be plural)

patrickangle commented on Source/WebCore/css/parser/CSSParserImpl.cpp:
```diff
@@ -592,6 +602,154 @@ RefPtr<StyleRuleFontFace> CSSParserImpl::consumeFontFaceRule(CSSParserTokenRange
+        m_observerWrapper->observer().startRuleHeader(StyleRuleType::FontFeatureValuesBlock, m_observerWrapper->startOffset(prelude));
+        m_observerWrapper->observer().endRuleHeader(m_observerWrapper->endOffset(prelude));
+        m_observerWrapper->observer().startRuleBody(m_observerWrapper->previousTokenStartOffset(range));
+        m_observerWrapper->observer().endRuleBody(m_observerWrapper->endOffset(range));
```
Most existing @ rules appear to call endRuleBody after they have consumed the range. It isn't immediately clear to me why or even if the end of the range can change during consumption, but it might be safer to follow that pattern here for consistency.

mdubet replied:
I've changed it to be extra-safe but I don't think it's necessary neither (FWIW, `consumeFontPalette` calls `endRuleBody` before consuming the range).

patrickangle commented on Source/WebCore/css/parser/CSSParserImpl.cpp:
```diff
@@ -592,6 +602,154 @@ RefPtr<StyleRuleFontFace> CSSParserImpl::consumeFontFaceRule(CSSParserTokenRange
+        m_observerWrapper->observer().startRuleHeader(StyleRuleType::FontFeatureValues, m_observerWrapper->startOffset(prelude));
+        m_observerWrapper->observer().endRuleHeader(m_observerWrapper->endOffset(prelude));
+        m_observerWrapper->observer().startRuleBody(m_observerWrapper->previousTokenStartOffset(block));
+        m_observerWrapper->observer().endRuleBody(m_observerWrapper->endOffset(block));
```
Ditto :639

patrickangle commented on Source/WebCore/css/parser/CSSParserImpl.cpp:
```diff
@@ -592,6 +602,154 @@ RefPtr<StyleRuleFontFace> CSSParserImpl::consumeFontFaceRule(CSSParserTokenRange
+        return nullptr;
+    
+    if (m_observerWrapper) {
+        m_observerWrapper->observer().startRuleHeader(StyleRuleType::FontFeatureValues, m_observerWrapper->startOffset(prelude));
```
We need to do this with a copy of prelude from before `consumeFamilyNameList`, otherwise the start offset and end offset of the prelude will be the same. It is still important this remain after the consumption though, since we don't want to inform the observer if the prelude wasn't parsable. See `CSSParserImpl::consumeLayerRule` for an example of that.

mdubet replied:
Fixed.

fujii commented on Source/WebCore/css/parser/CSSParserImpl.cpp:
```diff
@@ -593,6 +603,157 @@ RefPtr<StyleRuleFontFace> CSSParserImpl::consumeFontFaceRule(CSSParserTokenRange
+
+    auto [type, maxValues] = fontFeatureValuesTypeMappings(id);
+
+    auto consumeTag = [](CSSParserTokenRange range, uint maxValues) -> std::optional<Tag> {
```
AppleWin and WinCairo EWS reported a compilation error.
> C:\cygwin\home\buildbot\worker\Windows-EWS\build\Source\WebCore\css/parser/CSSParserImpl.cpp(645,53): error C2760: syntax error: 'identifier' was unexpected here; expected 'type specifier' (compiling source file C:\cygwin\home\buildbot\worker\Windows-EWS\build\WebKitBuild\Release\DerivedSources\WebCore\unified-sources\UnifiedSource-2f84417a-9.cpp) [C:\cygwin\home\buildbot\worker\Windows-EWS\build\WebKitBuild\Release\Source\WebCore\WebCore.vcxproj]

Replacing `uint` with `unsigned` solves the problem.

mdubet replied:
Thank you

mdubet committed:
Parsing for at-rule @font-feature-values
https://bugs.webkit.org/show_bug.cgi?id=246120
rdar://100830599

Reviewed by Myles C. Maxfield and Patrick Angle.

This implementation defines the @font-feature-values at-rule
and the six children font variation at-rules (@styleset, @stylistic, @character-variant,
@swash, @ornaments, @annotation), but those children at-rules are not exposed in CSSOM.

https://www.w3.org/TR/css-fonts-4/#font-feature-values-syntax

* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/test_font_feature_values_parsing-expected.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/JSCSSRuleCustom.cpp:
(WebCore::toJSNewlyCreated):
* Source/WebCore/css/CSSFontFeatureValuesRule.cpp: Added.
(WebCore::CSSFontFeatureValuesRule::CSSFontFeatureValuesRule):
(WebCore::CSSFontFeatureValuesRule::cssText const):
(WebCore::CSSFontFeatureValuesRule::reattach):
(WebCore::CSSFontFeatureValuesBlockRule::CSSFontFeatureValuesBlockRule):
(WebCore::CSSFontFeatureValuesBlockRule::cssText const):
(WebCore::CSSFontFeatureValuesBlockRule::reattach):
* Source/WebCore/css/CSSFontFeatureValuesRule.h: Added.
* Source/WebCore/css/CSSFontFeatureValuesRule.idl: Added.
* Source/WebCore/css/StyleRule.cpp:
(WebCore::StyleRuleBase::visitDerived):
(WebCore::StyleRuleBase::createCSSOMWrapper const):
(WebCore::StyleRuleFontFeatureValues::StyleRuleFontFeatureValues):
(WebCore::StyleRuleFontFeatureValuesBlock::StyleRuleFontFeatureValuesBlock):
(WebCore::StyleRuleFontFeatureValues::create):
* Source/WebCore/css/StyleRule.h:
(WebCore::StyleRuleBase::isFontFeatureValuesRule const):
(WebCore::StyleRuleBase::isFontFeatureValuesBlockRule const):
(isType):
* Source/WebCore/css/StyleRuleType.h:
* Source/WebCore/css/StyleSheetContents.cpp:
(WebCore::StyleSheetContents::traverseSubresources const):
* Source/WebCore/css/parser/CSSAtRuleID.cpp:
(WebCore::cssAtRuleID):
* Source/WebCore/css/parser/CSSAtRuleID.h:
* Source/WebCore/css/parser/CSSParserImpl.cpp:
(WebCore::computeNewAllowedRules):
(WebCore::CSSParserImpl::consumeRuleList):
(WebCore::CSSParserImpl::consumeAtRule):
(WebCore::CSSParserImpl::consumeQualifiedRule):
(WebCore::fontFeatureValuesTypeMappings):
(WebCore::CSSParserImpl::consumeFontFeatureValuesRuleBlock):
(WebCore::CSSParserImpl::consumeFontFeatureValuesRule):
* Source/WebCore/css/parser/CSSParserImpl.h:
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::consumeFamilyNameList):
* Source/WebCore/css/parser/CSSPropertyParserHelpers.h:
* Source/WebCore/inspector/InspectorStyleSheet.cpp:
(flattenSourceData):

Canonical link: https://commits.webkit.org/255677@main
